src/validations/html/rules/MismatchedTag.js

```diff
@@ -0,0 +1,139 @@
+import {List, Map, Record, Set} from 'immutable';
+import Code from '../rules/Code';
+
+const Location = Record({row: 0, column: 0});  // eslint-disable-line new-cap
+
+export default class {
+  constructor() {
+    this._openTagStack = List.of();
+    this._mismatchedTagErrorsByLocation = Map.of();
+    this._unopenedTagLocations = Set.of();
+    this._unopenedTagQueuesByName = Map.of();
+    this._foundSpareUnopenedTag = false;
+  }
+
+  openTag(location, {name}) {
+    this._openTagStack = this._openTagStack.push({location, name});
+  }
+
+  closeTag(location, name) {
+    const locationKey = new Location(location);
+
+    const openIndex = this._openTagStack.findLastIndex(
+      openTag => openTag.name === name);
+    const opened = openIndex >= 0;
+
+    // Assume any open tags after the matching open tag are mismatched. If the
+    // source is valid HTML, then no other open tags should exist. If no
+    // matching open tag exists, then all open tags are mismatched.
```

Comment:
Generally I’m against inline comments, as they tend to get out of sync with the actual code and I’m of the opinion that code can and should be written so that its intent is clear without additional comment. In this case, I think it’s quite clear what the next few lines of code are doing!